File: app/src/main/java/com/cookplan/recipe/import_recipe/parser/EdimDomaHtmlParser.java
package com.cookplan.recipe.import_recipe.parser;

import com.cookplan.R;
import com.cookplan.RApplication;
import com.cookplan.models.Ingredient;
import com.cookplan.models.MeasureUnit;
import com.cookplan.models.Product;
import com.cookplan.models.Recipe;
import com.cookplan.models.ShopListStatus;
import com.cookplan.providers.ProductProvider;
import com.cookplan.providers.impl.ProductProviderImpl;
import com.cookplan.utils.Utils;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by DariaEfimova on 09.06.17.
 */

public class EdimDomaHtmlParser extends BaseParser {

    private ProductProvider productDataProvider;
    private CompositeDisposable disposables;

    public EdimDomaHtmlParser(String url) {
        super(url);
        productDataProvider = new ProductProviderImpl();
        disposables = new CompositeDisposable();
    }

    @Override
    protected void parseDocument(Document doc) {
        List<String> names = getProductsNames(doc);
        disposables.add(
                productDataProvider.getTheClosestProductsToStrings(names)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new DisposableObserver<Map<String, List<Product>>>() {
                            @Override
                            public void onNext(Map<String, List<Product>> namesToProducts) {
                                disposables.clear();
                                Map<String, List<Ingredient>> ingredients = parceDocumentToIngredientList(doc, namesToProducts);
                                Recipe recipe = parceDocumentToRecipe(doc);
                                onImportSuccess(recipe, ingredients);

                            }


                            @Override
                            public void onError(Throwable e) {
                                onImportError(e);
                            }

                            @Override
                            public void onComplete() {

                            }
                        }));
    }

    private Recipe parceDocumentToRecipe(Document doc) {
        Recipe recipe = new Recipe();
        recipe.setName(parseRecipeTitleFromDoc(doc));

        String description = parseDescriptionFromDoc(doc);

        recipe.setDesc(description);

        List<String> imageUrls = parseImageUrlsFromDoc(doc);

        recipe.setImageUrls(imageUrls);
        return recipe;
    }


    private String parseDescriptionFromDoc(Document doc) {
        String description = "";
        Elements descriptionList = doc.select("div.content-box");
        for (Element desc : descriptionList) {
            if (desc.select("div.recipe-step-title").size() != 0) {
                description = description + desc.select("div.plain-text").text() + "\n";
            }
        }
        return description;
    }

    private List<String> parseImageUrlsFromDoc(Document doc) {
        List<String> imageUrls = new ArrayList<>();
        Elements imagesElements = doc.select("div.thumb-slider__slide");
        for (Element element : imagesElements) {
            if (element.children().size() == 1) {
                String url = element.child(0).attr("src");
                if (!url.isEmpty()) {
                    imageUrls.add("https:" + url);
                }
            }
        }
        return imageUrls;
    }

    private String getProductNameTag() {
        return "div.checkbox-info__name";
    }

    private String getIngredientItemTag() {
        return "tr.definition-list-table__tr";
    }

    private String getAmountTag() {
        return "td.definition-list-table__td_value";
    }

    private String parseRecipeTitleFromDoc(Document doc) {
        return doc.select("h1.recipe-header__name").text();
    }

    private List<String> getProductsNames(Document doc) {
        List<String> names = new ArrayList<>();

        Elements products = doc.select(getIngredientItemTag());
        for (Element product : products) {
            String name = parseStringProductName(product);
            if (!name.isEmpty()) {
                names.add(name);
            }
        }

        return names;
    }

    private String parseStringProductName(Element element) {
        Elements nameElem = element.select(getProductNameTag());
        String name = nameElem != null ? nameElem.text() : "";
        return name;
    }


    private String getMeasureUnitString(Double amount, Element element) {
        Elements amountElem = element.select(getAmountTag());
        if (amountElem.size() == 1) {
            String amountUnitString = amountElem.text();
            amountUnitString = amountUnitString.replace("½ ", "");
            amountUnitString = amountUnitString.replace("¾ ", "");
            String amountString;
            double value = Math.abs(amount - Math.round(amount));
            if (value < 1e-8 && value > -1e-8) {
                amountString = Integer.valueOf(amount.intValue()).toString();
            } else {
                amountString = amount.toString();
            }
            amountUnitString = amountUnitString.replace(amountString + " ", "");
            return amountUnitString;
        }
        return null;
    }

    private Map<String, List<Ingredient>> parceDocumentToIngredientList(Document doc, Map<String, List<Product>> namesToProducts) {
        Map<String, List<Ingredient>> ingredientMap = new HashMap<>();
        Elements ingredElements = doc.select(getIngredientItemTag());
        for (Element element : ingredElements) {
            String name = parseStringProductName(element);
            if (!name.isEmpty()) {
                List<Ingredient> ingredients = new ArrayList<>();
                for (Product product : namesToProducts.get(name)) {
                    ingredients.add(parseIngredient(product, element));
                }
                ingredientMap.put(name + ": " + element.select(getAmountTag()).text(), ingredients);
            }
        }
        return ingredientMap;
    }

    private Ingredient parseIngredient(Product product, Element element) {
        Double amount = parseAmount(element);
        MeasureUnit unit = parseMeasureUnit(amount, element);
        return new Ingredient(null,
                              product.toStringName(),
                              product,
                              null,
                              unit,
                              amount,
                              ShopListStatus.NONE);
    }

    private MeasureUnit parseMeasureUnit(Double amount, Element element) {
        MeasureUnit unit = MeasureUnit.UNITS;
        String unitString = getMeasureUnitString(amount, element);
        if (unitString != null) {
            unit = MeasureUnit.parseUnit(unitString);
        }
        return unit;
    }

    private Double parseAmount(Element element) {
        Double amount = 0.;
        Elements amountElem = element.select(getAmountTag());
        if (amountElem.size() == 1) {
            String amountString = amountElem.text();
            if (!amountString.contains(RApplication.Companion.getAppContext().getString(R.string.by_the_taste))) {
                String[] splited = amountString.split("\\s+");
                amount = Utils.getDoubleFromString(splited[0]);
            }
        }
        return amount;
    }
}

File: app/src/main/java/com/cookplan/recipe/import_recipe/parser/EdimDomaHtmlParser.kt
package com.cookplan.recipe.import_recipe.parser

import android.content.Context
import com.cookplan.models.*
import com.cookplan.providers.ProductProvider
import com.cookplan.providers.impl.ProductProviderImpl
import com.cookplan.utils.MeasureUnitUtils
import com.cookplan.utils.Utils
import com.google.firebase.auth.FirebaseAuth
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableObserver
import io.reactivex.schedulers.Schedulers
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element

/**
 * Created by DariaEfimova on 09.06.17.
 */

class EdimDomaHtmlParser(url: String, private val getContext: () -> Context?) : BaseParser(url) {

    private val productDataProvider: ProductProvider
    private val disposables: CompositeDisposable

    private val productNameTag: String
        get() = "div.checkbox-info__name"

    private val ingredientItemTag: String
        get() = "tr.definition-list-table__tr"

    private val amountTag: String
        get() = "td.definition-list-table__td_value"

    init {
        productDataProvider = ProductProviderImpl()
        disposables = CompositeDisposable()
    }

    override fun parseDocument(doc: Document) {
        val names = getProductsNames(doc)
        disposables.add(
                productDataProvider.getTheClosestProductsToStrings(names)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(object : DisposableObserver<MutableMap<String, List<Product>>>() {
                            override fun onNext(namesToProducts: MutableMap<String, List<Product>>) {
                                disposables.clear()
                                val ingredients = parceDocumentToIngredientList(doc, namesToProducts)
                                val recipe = parceDocumentToRecipe(doc)
                                onImportSuccess(recipe, ingredients)

                            }


                            override fun onError(e: Throwable) {
                                onImportError(e)
                            }

                            override fun onComplete() {

                            }
                        }))
    }

    private fun parceDocumentToRecipe(doc: Document): Recipe {
        val name = parseRecipeTitleFromDoc(doc)

        val description = parseDescriptionFromDoc(doc)

        val imageUrls = parseImageUrlsFromDoc(doc)

        return Recipe(
                id = String(),
                name = name,
                desc = description,
                descImageUrls = arrayListOf(),
                imageUrls = imageUrls,
                userId = FirebaseAuth.getInstance().currentUser?.uid,
                userName = FirebaseAuth.getInstance().currentUser?.displayName)
    }


    private fun parseDescriptionFromDoc(doc: Document): String {
        var description = String()
        val descriptionList = doc.select("div.content-box")
        descriptionList
                .asSequence()
                .filter { it.select("div.recipe-step-title").size != 0 }
                .forEach { description = description + it.select("div.plain-text").text() + "\n" }
        return description
    }

    private fun parseImageUrlsFromDoc(doc: Document): List<String> {
        val imageUrls = mutableListOf<String>()
        val imagesElements = doc.select("div.thumb-slider__slide")
        imagesElements
                .filter { it.children().size == 1 }
                .map { it.child(0).attr("src") }
                .filterNot { it.isEmpty() }
                .mapTo(imageUrls) { "https:" + it }
        return imageUrls
    }

    private fun parseRecipeTitleFromDoc(doc: Document): String {
        return doc.select("h1.recipe-header__name").text()
    }

    private fun getProductsNames(doc: Document): List<String> {
        val products = doc.select(ingredientItemTag)
        val names = products
                .map { parseStringProductName(it) }
                .filterNot { it.isEmpty() }

        return names
    }

    private fun parseStringProductName(element: Element): String {
        val nameElem = element.select(productNameTag)
        return nameElem.text() ?: String()
    }


    private fun getMeasureUnitString(amount: Double, element: Element): String {
        val amountElem = element.select(amountTag)
        var amountUnitString = String()
        if (amountElem.size == 1) {
            amountUnitString = amountElem.text()
            amountUnitString = amountUnitString.replace("½ ", "")
            amountUnitString = amountUnitString.replace("¾ ", "")
            val amountString: String
            val value = Math.abs(amount - Math.round(amount))
            if (value < 1e-8 && value > -1e-8) {
                amountString = amount.toInt().toString()
            } else {
                amountString = amount.toString()
            }
            amountUnitString = amountUnitString.replace(amountString + " ", "")
        }
        return amountUnitString
    }

    private fun parceDocumentToIngredientList(doc: Document, namesToProducts: MutableMap<String, List<Product>>): MutableMap<String, List<Ingredient>> {
        val ingredientMap = mutableMapOf<String, List<Ingredient>>()
        val ingredElements = doc.select(ingredientItemTag)
        for (element in ingredElements) {
            val name = parseStringProductName(element)
            if (!name.isEmpty()) {
                val ingredients = (namesToProducts[name] ?: listOf()).map { parseIngredient(it, element) }
                ingredientMap.put(name + ": " + element.select(amountTag).text(), ingredients)
            }
        }
        return ingredientMap
    }

    private fun parseIngredient(product: Product, element: Element): Ingredient {
        val amount = parseAmount(element)
        val unit = parseMeasureUnit(amount, element)
        return Ingredient(null,
                product.toStringName(),
                product, null,
                unit,
                amount,
                ShopListStatus.NONE)
    }

    private fun parseMeasureUnit(amount: Double, element: Element): MeasureUnit =
            MeasureUnitUtils.parseUnit(getMeasureUnitString(amount, element), getContext)

    private fun parseAmount(element: Element): Double {
        var amount = 0.toDouble()
        val amountElem = element.select(amountTag)
        if (amountElem.size == 1) {
            val amountString = amountElem.text()
            if (!amountString.contains(MeasureUnit.Companion.getByTasteString())) {
                val splited = amountString.split("\\s+".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                amount = Utils.getDoubleFromString(splited[0])
            }
        }
        return amount
    }
}